pkg/config/user_config.go

```diff
@@ -478,3 +482,17 @@ func GetDefaultConfig() *UserConfig {
 		NotARepository:       "prompt",
 	}
 }
+
+func getPullModeFromGitConfig() string {
+	rebaseOut, rebaseErr := secureexec.Command("git config --get pull.rebase").Output()
```

Comment:
I wonder if this is the right place to be fetching this information. At the moment we have out GitCommand struct in `pkg/commands/git.go` being passed a `getGitConfigValue` function (whose default implementation is defined in `pkg/commands/gitconfig.go`), and when we need to get a value we just do it directly rather than setting the config value at startup. Doing it that way helps with testing where we can pass a mock `getGitConfigValue` when instantiating the GitCommand struct.

That said, I'm open to getting all this stuff at startup so save on re-fetching it each time we want to pull, but I think if we do that we should apply it across the board. There is value to fetching it each time we need it though because doing so is very fast (iirc) and allows the config to be updated on the fly without needing to restart lazygit.

What are your thoughts?

Reply:
I may be mistaken, but from what I've seen in the code the configuration is only loaded once when the app is being initialized (to be fair it's loaded twice, idk if this is intentional) and stored in an `AppConfig` class which is later accessed by everything else.
I've also seen the `git.go` package but the constructor requires so much stuff: a logger, an os command executor, translations, and app configurer that would make the code a mess.
I have to admit that I missed the `gitconfig.go`, I could use that.

Here's where I found the configuration is being loaded: 
 - In `NewAppConfig`, only called once in main and stored globally
https://github.com/jesseduffield/lazygit/blob/e42387d0daa45a48666eb837540ce871bb71fd8e/pkg/config/app_config.go#L54

 - In `gui.Run()`, called by `gui.RunAndHandleError`, called by `app.Run()`, again, only called in main and stored globally
  https://github.com/jesseduffield/lazygit/blob/e42387d0daa45a48666eb837540ce871bb71fd8e/pkg/gui/gui.go#L519

Reply:
Ah, the ReloadUserConfig thing is now outdated: previously whenever switching to a subprocess we would stop the gui and then start it again upon returning. On the back of a request to reload the gui when returning, I chucked that in the Run method, but now that we're only calling Run once and we're using tcell's suspend/resume API, we should reload the user config whenever returning from a subprocess (outside the scope of this PR).

I just re-measured how long these git config commands take. Looks like around 7ms, which I think is too long to be doing them on the fly. We wouldn't want to do them sequentially at startup though because I think it's important to keep startup as fast as possible. We can use a waitgroup and goroutines to get around that i.e. spawn a separate goroutine for each git config variable we need.

We're currently fetching git config things via the `GetConfigValue` method
![image](https://user-images.githubusercontent.com/8456633/114687149-ca32b080-9d56-11eb-9dc4-b9dd9fda1403.png)

So I'm thinking we should either get all those things happening when getting the user config, using waitgroups to do it efficiently, or we continue using the existing pattern for now. What are your thoughts?

Reply:
It takes a while for the user to open the application and execute the pull command. I guess It's possible to use a background task to load this and prevent any functions from doing a pull until the configuration is fetched.
Speaking about updating configuration on the fly, does this app benefit from that? I've never had it open for longer than a minute and I believe the most common workflow is open -> do stuff -> close.

My proposal is the following:
 - Replace the secureexec with some getGitConfigValue in a way it can be mocked and tested
 - Add test
 - Load pull configuration in a non-blocking way (only block pull tasks)

Reply:
@eruizc-dev that works for me